### PR TITLE
Walk

### PR DESCRIPTION
**Pull request by alice**

### CODE REVIEW

File: inert.js [hunk 1]
```diff
@@ -15,76 +15,307 @@
...
+function watchForInert(records, self) {
   for (var record of records) {
     if (record.type != 'attributes')
       continue;
     if (record.attributeName != 'inert')
       continue;
     var target = record.target;
-    if (target.detabinator === undefined)
-        target.detabinator = new Detabinator(target);
     var inert = target.hasAttribute('inert');
```
Comment (alice) [inline inert.js]: could this be: if (record.type == 'attributes' && record.attributeName == 'inert')? (from @robdodson)

File: inert.js [hunk 2]
```diff
@@ -15,76 +15,307 @@
...
-    if (!element) {
-      throw new Error('Missing required argument. new Detabinator needs an element reference');
+class InertRoot {
+  constructor(rootElement, inertManager) {
+    this._inertManager = inertManager;
+    this._rootElement = rootElement;
+    this._rootElement.setAttribute('aria-hidden', 'true');
+    this._rootElement.__inertRoot = this;
+
+    this._savedNodes = new Set([]);
```
Comment (jessebeach) [inline inert.js]: You never access the savedNodes set; just additions and deletions. I'm guessing you're usingteh manager to track the nodes? Maybe you can simply delete this private member.

File: inert.js [hunk 3]
```diff
@@ -15,76 +15,307 @@
...
+   */
+  constructor(document) {
+    if (!document)
+      throw new Error('Missing required argument; InertManager needs to wrap a document.');
+
+    this._document = document;
+    this._savedNodes = new Set([]);
+    this._inertRoots = new Set([]);
+
+    var inertElements = Array.from(document.querySelectorAll('[inert]'));
```
Comment (alice) [inline inert.js]: might consider switching all uses of var to const unless they're mutable and then keeping them let
(from @robdodson)

File: inert.js [hunk 4]
```diff
@@ -15,76 +15,307 @@
...
+      this.setInert(inertElement, true);
+  }
+
+  /**
+   * @param {Element} root
+   * @param {boolean} inert
+   */
+  setInert(root, inert) {
+    if (inert) {
+      if (root.__inertRoot)  // Check for inert attribute?
```
Comment (alice) [inline inert.js]: maybe add comments to 187 and 195 explicitly saying 'if element is already {inert|not inert} then return' (from@robdodson)

File: inert.js [hunk 5]
```diff
@@ -15,76 +15,307 @@
...
 
-  set inert(isInert) {
-    if (this._inert === isInert) {
+  /**
+   * @param {Node} node
+   */
+  _visitNode(node) {
+    if (node.hasAttribute('inert')) {
+      let inertSubRoot = node.__inertRoot;
+      for (let savedInertNode of inertSubRoot._savedNodes) {
```
Comment (jessebeach) [inline inert.js]: Actually, I guess you loop through the savedNodes here. Can you just use the manager's set of nodes? It seems danger to track to sets of nodes that might get out of sync.

File: inert.js [hunk 6]
```diff
@@ -15,76 +15,307 @@
...
+class InertRoot {
+  constructor(rootElement, inertManager) {
+    this._inertManager = inertManager;
+    this._rootElement = rootElement;
+    this._rootElement.setAttribute('aria-hidden', 'true');
+    this._rootElement.__inertRoot = this;
+
+    this._savedNodes = new Set([]);
+
+    this._walkSubtree(this._rootElement);
```
Comment (alice) [inline inert.js]: Maybe comment what's happening here;  walkSubtree as a method name doesn't necessarily imply the side effects that will occur (from @robdodson)

File: inert.js [hunk 7]
```diff
@@ -15,76 +15,307 @@
...
+    InertManager.composedTreeWalk(startNode, { preorder: (node) => { return this._visitNode(node); } });
   }
 
-  set inert(isInert) {
-    if (this._inert === isInert) {
+  /**
+   * @param {Node} node
+   */
+  _visitNode(node) {
+    if (node.hasAttribute('inert')) {
```
Comment (alice) [inline inert.js]: Check not visiting self here

File: inert.js [hunk 8]
```diff
@@ -15,76 +15,307 @@
...
+    if (!node.matches(InertNode._focusableElementsString) && !node.hasAttribute('tabindex'))
       return;
+
+    let inertNode = this._inertManager.register(node, this);
+    this._savedNodes.add(inertNode);
+  }
+
+  _onChildAdded(records, self) {
+    for (let record of records) {
+      for (let node of Array.from(record.addedNodes))
```
Comment (jessebeach) [inline inert.js]: Where are addedNodes added? I can't find an add expression.

File: inert.js (hunk 5 above)
Comment (alice) [inline inert.js]: The `InertManager` has the list of all managed nodes. Each `InertRoot` knows about *its* managed nodes. This is important when e.g. we want to remove an `InertRoot` below another `InertRoot`.

File: inert.js [hunk 9]
```diff
@@ -15,76 +15,307 @@
...
+    }
+    delete this._node.__inertNode;
+    delete this._node;
+    delete this._inertRoots;
+
+    this._destroyed = true;
+  }
+
+  static get _focusableElementsString() {
+    return 'a[href],area[href],input:not([disabled]),select:not([disabled]),textarea:not([disabled]),button:not([disabled]),iframe,object,embed,[contenteditable]';
```
Comment (jessebeach) [inline inert.js]: nit: use [].join(',') and put each string on it's own line for readability.

File: inert.js (hunk 8 above)
Comment (alice) [inline inert.js]: It is part of the MutationObserver API.

File: inert.js [hunk 10]
```diff
@@ -15,76 +15,307 @@
...
+    } else if (node.hasAttribute('tabindex')) {
+      this._savedTabIndex = node.tabIndex;
+      this._node.removeAttribute('tabindex');
+    }
+  }
+
+  destructor() {
+    this._throwIfDestroyed();
+
+    if (this._node.parentNode) {
```
Comment (alice) [inline inert.js]: why do you check for a parentNode here? (from @robdodson)

Comment (alice) [inline inert.js]: In case the node has been detached (though maybe this should check for `ownerDocument` instead)

File: inert.js [hunk 11]
```diff
@@ -15,76 +15,307 @@
...
+  }
+
+  /**
+   * @param {Node} node
+   * @param {InertRoot} inertRoot
+   * @return {InertNode} inertNode
+   */
+  register(node, inertRoot) {
+    let inertNode = undefined;
+    // TODO better way to associate nodeData with node?
```
Comment (jessebeach) [inline inert.js]: What about a Map in the manager keyed by node?

File: inert.js [hunk 12]
```diff
@@ -15,76 +15,307 @@
...
+    }
+
+    if (element && callbacks.postorder)
+      callbacks.postorder.call(null, element);
   }
+
 }
 
-function observe(records, self) {
+document.inertManager = new InertManager(document);
```
Comment (jessebeach) [inline inert.js]: Why not just store this is a scoped var in the module rather than modifying the document?

File: inert.js [hunk 13]
```diff
@@ -15,76 +15,307 @@
...
     var target = record.target;
-    if (target.detabinator === undefined)
-        target.detabinator = new Detabinator(target);
     var inert = target.hasAttribute('inert');
-    target.detabinator.inert = inert;
-    if (inert)
-      target.setAttribute('aria-hidden', 'true');
-    else
-      target.removeAttribute('aria-hidden')
+    document.inertManager.setInert(target, inert);
```
Comment (jessebeach) [inline inert.js]: ditto private module var from above

File: inert.js [hunk 14]
```diff
@@ -15,76 +15,307 @@
...
+    var child = node.firstChild;
+    while (child != null) {
+      InertManager.composedTreeWalk(child,
+                                    callbacks,
+                                    opt_shadowRoot);
+      child = child.nextSibling;
+    }
+
+    if (element && callbacks.postorder)
+      callbacks.postorder.call(null, element);
```
Comment (jessebeach) [inline inert.js]: nit: invoke directly; not need to muck with invocation context.

File: inert.js (hunk 12 above)
Comment (alice) [inline inert.js]: This is the public API effectively, so putting it on the document makes it easy to find and use.

File: inert.js [hunk 15]
```diff
@@ -15,76 +15,307 @@
...
+  _visitNode(node) {
+    if (node.hasAttribute('inert')) {
+      let inertSubRoot = node.__inertRoot;
+      for (let savedInertNode of inertSubRoot._savedNodes) {
+        this._inertManager.register(savedInertNode.node, this);
+        this._savedNodes.add(savedInertNode);
+      }
+    }
+
+    if (!node.matches(InertNode._focusableElementsString) && !node.hasAttribute('tabindex'))
```
Comment (robdodson) [inline inert.js]: nit: feels a little funky to access an underscored property like this. Should the underscore just be removed?

Comment (robdodson) [inline inert.js]: I guess you could also wrap everything in an IIFE and make it a `const`. Doesn't feel like it really belongs to anyone

File: inert.js [hunk 16]
```diff
@@ -15,76 +15,307 @@
...
+    delete this._inertManager;
   }
 
-  get inert() {
-    return this._inert;
+  /**
+   * @param {Node} startNode
+   */
+  _walkSubtree(startNode) {
+    InertManager.composedTreeWalk(startNode, { preorder: (node) => { return this._visitNode(node); } });
```
Comment (robdodson) [inline inert.js]: the `preorder` callback here will never return `true` correct?

Comment (robdodson) [inline inert.js]: if that's correct then it seems weird to have all the preorder/postorder machinery.

File: inert.js (hunk 15 above)
Comment (robdodson) [inline inert.js]: Also, could this be changed to `if (node.matches(InertNode._focusableElementsString) || node.hasAttribute('tabindex'))` and then take the next couple of lines and place them in the conditional? I kind of glossed over this chunk a few times. Maybe it's just the way my brain is working today but I find it easier for the code to be explicit about what it's looking for

File: inert.js (hunk 7 above)
Comment (robdodson) [inline inert.js]: Good place to add a comment that you're specifically checking for nested inertRoots and why you're doing that

File: inert.js (hunk 16 above)
Comment (alice) [inline inert.js]: Yeah that's there cause it was necessary in the previous context for that code :)

File: inert.js [hunk 17]
```diff
@@ -15,76 +15,307 @@
...
+
+  /**
+   * @param {Node} node
+   * @param {InertRoot} inertRoot
+   * @return {InertNode} inertNode
+   */
+  register(node, inertRoot) {
+    let inertNode = undefined;
+    // TODO better way to associate nodeData with node?
+    if (node.__inertNode) {
```
Comment (robdodson) [inline inert.js]: In what scenario would we go down this code path?

File: inert.js (hunk 1 above)
Comment (alice) [inline inert.js]: Prefer to early out but combined the two continues into one.

File: inert.js (hunk 2 above)
Comment (alice) [inline inert.js]: As discussed, got rid of the per-node state instead and made a Map.

File: inert.js (hunk 3 above)
Comment (alice) [inline inert.js]: I got rid of most `var` but I'm still getting my head around what `const` means in JS :confused:

File: inert.js (hunk 4 above)
Comment (alice) [inline inert.js]: Done.

File: inert.js (hunk 6 above)
Comment (alice) [inline inert.js]: Changed name to `_makeSubtreeUnfocusable`.

File: inert.js (hunk 7 above)
Comment (alice) [inline inert.js]: Done.

File: inert.js (hunk 9 above)
Comment (alice) [inline inert.js]: Done.

File: inert.js (hunk 10 above)
Comment (alice) [inline inert.js]: Removed. If someone wants to un-inert a detached element, more power to 'em.

File: inert.js (hunk 11 above)
Comment (alice) [inline inert.js]: Done.

File: inert.js (hunk 12 above)
Comment (alice) [inline inert.js]: Then I changed the API so now it's scoped :)

File: inert.js (hunk 13 above)
Comment (alice) [inline inert.js]: Ditto.

File: inert.js (hunk 14 above)
Comment (alice) [inline inert.js]: Obsolete comment.

File: inert.js (hunk 15 above)
Comment (alice) [inline inert.js]: Done and done.

(I really don't like when methods end with the end of an if block for some reason, but I'm working on it.)

File: inert.js (hunk 16 above)
Comment (alice) [inline inert.js]: Simplified.

File: inert.js (hunk 17 above)
Comment (alice) [inline inert.js]: Added comment:
```js
// This node was already in an inert subtree
```

File: inert.js [hunk 18]
```diff
@@ -15,76 +15,416 @@
...
+    this._observer = new MutationObserver(this._onChildAdded.bind(this));
+    this._observer.observe(this._rootElement, { childList: true, subtree: true });
+  }
+
+  /**
+   * Call this whenever this object is about to become obsolete.
+   * This unwinds all of the state stored in this object and updates the state of all of the managed nodes.
+   */
+  destructor() {
+    console.log('destroying', this);
```
Comment (jessebeach) [inline inert.js]: These console calls should be pulled out.

Comment (alice) [inline inert.js]: Oh thanks!

Comment (alice) [inline inert.js]: Done.

**Pull request merged**